### DISCUSSION TITLE
What is the purpose of Language API?

### DISCUSSION THREAD

**Discussion by JeganRavichandran** [microsoft/WindowsAppSDK#1256] [Q&A] [answered]
I am a control developer, I want my control to support location, globalization. I want my control to follow Microsoft's way instead of solving using explicit APIs.

There are some APIs to change the culture of an application.

1. Thread.CurrentThread.CurrentCulture
2. Thread.CurrentThread.CurrentUICulture
3. ApplicationLanguages.PrimaryLanguageOverride
4. FrameworkElement.Language

Question:
1. We were using CurrentCulture & CurrentUICulture for so long. What is the purpose of PrimaryLanguageOverride and Language API. When to use it?
2. Is it possible to change the culture for specific control?  that is, I have 2 instances of the same control, I want both these instances to have a different culture. Can I use Language API for this purpose?
3. Based on which API I should choose a resource (resw) for localization?
4. What is the standard way to change the localization at runtime?

**Comment by evelynwu-msft** [comment, marked as answer]
I recommend familiarizing yourself with this [documentation](https://docs.microsoft.com/en-us/windows/uwp/app-resources/) on MSDN that goes over the resource management system used for Windows apps, and in particular this [section](https://docs.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest) on how to provide localized strings.

1. `System.Threading.Thread.CurrentCulture/CurrentUICulture` are .NET concepts. I'm not familiar with their implementation so I can't say just how far down the stack they go (i.e. do they apply only to .NET code or would they also affect native code in the process like WinUI). MSDN has pretty thorough documentation on both [ApplicationLanguages.PrimaryLanguageOverride](https://docs.microsoft.com/en-us/uwp/api/windows.globalization.applicationlanguages.primarylanguageoverride?view=winrt-20348) and [FrameworkElement.Language](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.language?view=winrt-20348). It should be noted that the latter only affects text *formatting* e.g. font fallback.
2. No.
3. `.resw` is currently the only localization method supported by WinUI (i.e. both the framework and the app will have consistent locale behavior), although we have plans to introduce in a future release the ability to plug in other localization methods (e.g. .NET's `System.Resources.ResourceManager`).
4. `ApplicationLanguages.PrimaryLanguageOverride`; this needs to be set fairly early during app startup (e.g. the `App` class's constructor) in order to have an effect. But it is not recommended that you do this; instead, you should leave it to the user to decide (in their OS settings) what language should be used.

**Reply by JeganRavichandran** [reply]
I am not able to mark your response as an answer.

**Reply by oliverw** [reply]
> 4. `ApplicationLanguages.PrimaryLanguageOverride`; this needs to be set fairly early during app startup (e.g. the `App` class's constructor) in order to have an effect. But it is not recommended that you do this; instead, you should leave it to the user to decide (in their OS settings) what language should be used.

There are lots of users working in a language environment they do not control and which are grateful to be given the possibility to override the OS defaults. Besides `ApplicationLanguages.PrimaryLanguageOverride` does not even work reliably as it currently take two application restarts to have the changed value apply.

**Reply by DarranRowe** [reply]
Even worse, it is only supported on applications that have a package identity.
